### PR TITLE
blog: Exp versioning

### PR DESCRIPTION
**Pull request by dberenbaum**
First draft of experiment versioning blog post. Mostly worked on the intro. The tutorial section is outdated, mainly from when we were still focusing more on "lightweight experiments" rather than "experiment versioning."

### CODE REVIEW

**Comment by dberenbaum** [comment]
Hm, not sure why this isn't deploying. Hoping to start sharing it around for feedback if I can get a preview deployed.

**Comment by shcheklein** [comment]
I've requested the redeploy. @dberenbaum does it work for you locally?

@rogermparent I wonder if there is a way to connect Heroku better - see the logs here and a bot that would react to `/rebuild` command or something.

**Comment by dberenbaum** [comment]
Sorry, I should have debugged locally first! It's working locally now, but I can't figure out how to add myself as an author. I tried adding myself to `content/authors` but still getting an error. Also not sure where/how to include the avatar since I see them in multiple folders, and I'm not sure if there are size requirements or anything.

**Comment by shcheklein** [comment]
@julieg18 @rogermparent could we please help Dave here and update the contrib docs if needed to help other folks in the future?

**Comment by dberenbaum** [comment]
Ah, sorry again! I didn't realize we had https://dvc.org/doc/user-guide/contributing/blog. Let me try once more.

**Comment by dberenbaum** [comment]
Okay, got it working, thanks!

**Comment by dberenbaum** [comment]
This is still not ready for copy editing, but any other feedback would be appreciated!

**Comment by jendefig** [comment]
Only other thing that needs to be poked in somewhere is link to command ref doc. Maybe in more than one place.  Where appropriate in body and also at the end.

**Comment by dberenbaum** [comment]
> Only other thing that needs to be poked in somewhere is link to command ref doc. Maybe in more than one place. Where appropriate in body and also at the end.

What other links would be useful to add @jendefig?

**Comment by jendefig** [comment]
> What other links would be useful to add @jendefig?

Well we discussed putting the video in.  I think after the first paragraph would be a good place.  Then they can continue to read in more detail.  

The other links that are in there are good, but just need the right formatting.

The only other things I can think of is to direct to Jorge's Use case doc that should be up and maybe (?) to the meetup where we will have demo (Join us at our December Office Hours for a demo on `dvc exp init` ! ) But that will be tricky because I will have to have the event posted in Meetup already.  But I can do that by doing some kind of "we have something really exciting planned this day, but I can't tell you yet" kind of thing and then change the info to Meetup correctly on the day of the release.  That would work. And potentially build anticipation.

File: content/blog/2021-11-17-exp-versioning.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+commentsUrl:
+tags:
+  - MLOps
+  - DVC
+  - Experiments
+  - Experiment Tracking
+  - Experiment Versioning
+---
+
+Traditional software version control is not enough for tracking machine learning
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:23]: Good and strong statement. I wish we have something similar in the description and probably in the title.

File: content/blog/2021-11-17-exp-versioning.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+experiments. Over the last few years, experiment tracking has come a long way,
+but there are parallels to problems from earlier generations of version control.
+That's why "experiment versioning" in DVC tracks experiments by building on
+modern version control practices.
+
+# Experiment Tracking Needs
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:29]: This paragraph is good but a bit too abstract for the 1st one. I'd suggest to extend it by some examples/pictures or remove/move-for-later.

File: content/blog/2021-11-17-exp-versioning.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+│
+│ ├── 9a4ff1c [exp-333c9] │ 10:40 AM     │ 0.25183 │ 0.9137 │ 10           │ 64
+│
+│ ├── 138e6ea [exp-55e90] │ 10:28 AM     │ 0.25784 │ 0.9084 │ 10           │ 32
+│
+│ └── 51b0324 [exp-2b728] │ 10:17 AM     │ 0.25829 │ 0.9058 │ 10           │ 16
+│
+└─────────────────────────┴──────────────┴─────────┴────────┴──────────────┴──────────────────┘
+
+$ dvc exp branch exp-333c9 conv-units-64
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:205]: This might look better in a separate "terminal" or even a separate (section).

File: content/blog/2021-11-17-exp-versioning.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        git checkout conv-units-64
+
+```
+
+Reproducing experiments can be automated since your experiment pipeline is
+codified, and DVC will check for cached inputs and outputs and skip recomputing
+data that's been generated before.
+
+```dvc
+$ dvc exp run
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:218]: Value of this step might be not clear. Should user promote this exp or continue working on this idea - run next exp?

File: content/blog/2021-11-17-exp-versioning.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+# What Next?
+
+Experiment versioning is still in its early days. Look out for future
+announcements about:
+
+- Deep learning challenges like live monitoring, checkpointing, and iterative
+  training.
+- Visualizing and comparing experiment results in other tools like VSCode and
+  Studio.
+- Orchestrating experiments on other machines.
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:239]: It is better to have a more focused/short conclusion. 3 items might be a bit too overwhelming. 1 should be enough.

File: content/blog/2021-11-17-exp-versioning.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+What do you want to see for the next generation of experiment tracking? Try out
+experiment versioning, share your thoughts, and help build the future of ML
+developer tools.
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:243]: It feels like call to action but without action :) What I'm suppose to do? Comment here, provide a feedback on Discord, tweet with mentioning @DVCOrg?

Or it might be ok to keep it as is 🤷

File: content/blog/2021-11-17-exp-versioning.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+See also https://git-scm.com/book/en/v2/Getting-Started-About-Version-Control.
+
+Now let's compare to experiment tracking:
+
+| Generation | Platform                   | Logging        | Organization | Reproducibility | Collaboration |
+| ---------- | -------------------------- | -------------- | ------------ | --------------- | ------------- |
+| First      | Spreadsheet, notebook      | Scalar values  | None         | None            | None          |
+| Second     | API + database + dashboard | Multi-artifact | Labels       | Manual          | Centralized   |
+| Third      | DVC + Git                  | Changesets     | Git-based    | Automatic       | Distributed   |
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:102]: `DVC + Git` - there are no other names but DVC. It might be better to include other names or not to use DVC.
Git is fine I think.

**Review by dmpetrov — CHANGES_REQUESTED** [review]
Great blog post, @dberenbaum! Very methodical with a clear structure: problem --> history --> solution with example.

It is good (and probably the best from technical point of view) but might be too deep\long for a wide audience. It lacks real life examples. I guess, a good portion of our audience values the problem Exp Versioning solves more then machinery behind it including definition of experiments and history. I'd try to create the entire post (or the first part of it) based on examples (what I cannot properly do with MLFlow but I can with DVC?) and talk about experiment definitions and history in the 2nd part. The history is a good foundation for strong conclusion.

However, I'm afraid that this redesign might break a well structured blog. We should think about the strategy - trying to fit a wider audience of users by a redesign or create a separate blog post for a wider audience and keep this in more engineering style.

@dberenbaum what do you think?

File: content/blog/2021-11-17-exp-versioning.md (hunk 6 above)
Comment (MKhalusova) [reply content/blog/2021-11-17-exp-versioning.md:243]: It can be as simple as "What do you want to see for the next generation of experiment tracking? Let us know in the comments!"

File: content/blog/2021-11-17-exp-versioning.md (hunk 5 above)
Comment (MKhalusova) [reply content/blog/2021-11-17-exp-versioning.md:239]: @dmpetrov I disagree about reducing the list to 1 item. When reading a "what's next" section I typically expect more than 1 thing. And 3 is ideal, as more than that can get overwhelming.
However, it feels like the announcements should not be about challenges, but about "new ways to overcome challenges", about the positive value that we provide.

File: content/blog/2021-11-17-exp-versioning.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+```
+
+# What Next?
+
+Experiment versioning is still in its early days. Look out for future
+announcements about:
+
+- Deep learning challenges like live monitoring, checkpointing, and iterative
+  training.
+- Visualizing and comparing experiment results in other tools like VSCode and
```
Comment (MKhalusova) [inline content/blog/2021-11-17-exp-versioning.md:237]: @dberenbaum I believe, it's abbreviated to VS Code, with a space between VS and Code. 
Also, I am not sure how many people know Studio, and personally, I would refer to it as "Studio by iterative.ai", and add a link to its homepage for those who are curious.

File: content/blog/2021-11-17-exp-versioning.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+- Log experiment information.
+- Organize and compare experiments.
+- Reproduce experiments.
+
+Tracking data, comparing versions, and enabling reproducibility are already
+addressed by version control. Unfortunately, traditional version control tools
+like Git don't work well for ML experimentation because:
+
+- Data and models aren't easily tracked with Git.
+- There are too many experiments to track with any branching workflow.
```
Comment (MKhalusova) [inline content/blog/2021-11-17-exp-versioning.md:42]: Probably a silly question, but we still use branches for tracking experiments with DVC, don't we?  If so, the problem of "too many branches" is still there

File: content/blog/2021-11-17-exp-versioning.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+# Experiment Versioning in DVC
+
+How does DVC help?
+
+- Log experiments end to end (commands, inputs, and outputs) and track changes
+  introduced by each experiment.
+- Associate experiments with Git commits, branches, and tags from which they
+  derive. Organize locally and choose what to share.
+- Reproduce any previous experiment with a single command. Cache previous runs
+  and results to reduce duplicate compute time and storage space.
```
Comment (MKhalusova) [inline content/blog/2021-11-17-exp-versioning.md:82]: This may be me, but it would help to have a simple example of the difference. Let's say. Here's a problem. When we solve it with some other tool (not mentioning), this is what we can log and track. 
When we solve it with DVC, these are the additional benefits.

**Comment by dberenbaum** [comment]
> It is good (and probably the best from technical point of view) but might be too deep\long for a wide audience. It lacks real life examples. I guess, a good portion of our audience values the problem Exp Versioning solves more then machinery behind it including definition of experiments and history. I'd try to create the entire post (or the first part of it) based on examples (what I cannot properly do with MLFlow but I can with DVC?) and talk about experiment definitions and history in the 2nd part. The history is a good foundation for strong conclusion.

Yes, agree this is the key question. I want to wait on addressing the smaller comments until we get a good basic structure in place. I had the same thought that the history stuff might be too much buildup that users don't really care about, depending on what kind of post we want this to be. I had already considered that a lot of the history might get cut out, so no big deal if that's what happens.

Let's discuss this week. I've been working through some ideas that I can also show a little bit during the DVC working session or elsewhere. One challenge is that the biggest advantages shine through in complex examples (working on multiple branches, going back in history after developing many experiments, etc.), so it can be hard to develop succinct examples. Especially if we also want to highlight `dvc exp init` and getting started quickly, which focuses on more simple use cases.

File: content/blog/2021-11-17-exp-versioning.md (hunk 9 above)
Comment (iesahin) [reply content/blog/2021-11-17-exp-versioning.md:42]: `dvc exp` uses Git references in the form `refs/exps/123ABCD/exp-4484a` for experiments. These are outside of the branch structure, but can be moved to Git via `dvc exp branch`. The user doesn't think of a way to create branches for experiments with DVC, selects them for branches afterwards.

Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:42]: We sometimes call them lightweight branches or hidden branches. The point is that the user does not need to manually generate commits nor branches, nor do they see those experiments when using typical git commands like `git branch`, and there are commands like `dvc exp remove` to clean them up.

File: content/blog/2021-11-17-exp-versioning.md (hunk 6 above)
Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:243]: Good suggestion, @MKhalusova! I used that for now, but we might want to discuss to also link to other channels here (twitter, discord, etc.).

**Comment by dberenbaum** [comment]
@dmpetrov and others, PTAL at the latest draft when you have a chance. Thanks!

**Comment by jendefig** [comment]
@dberenbaum Here is the link to the  December OH meetup in which we will have the demo: https://www.meetup.com/DVC-Community-Virtual-Meetups/events/282064369/

File: content/blog/2021-11-17-exp-versioning.md [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+```dvc
+$ dvc exp branch exp-333c9 conv-units-64
+Git branch 'conv-units-64' has been created from experiment 'exp-333c9'.
+To switch to the new branch run:
+
+        git checkout conv-units-64
+
+```
+
+# History of Experiment Tracking
```
Comment (dmpetrov) [inline content/blog/2021-11-17-exp-versioning.md:262]: That's really nice analogy. However, it might be better to focus on functionality and keep it for a next blog post. If you decide to do that it still worth to keep a paragraph to reference the history.

**Review by dmpetrov — APPROVED** [review]
@dberenbaum great blog post. Definitely the best we have about experiments!

The major question - should we keep it as tutorial or polish it a bit to make it to more "viral" style (not necessary reproducible). My opinion - it worth to continue polish it a bit.

File: content/blog/2021-11-17-exp-versioning.md [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+
+Traditional software version control is not enough for tracking machine learning
+experiments. When tracking ML experiments, you need to not only version code,
+but to:
+
+- Log experiment information (like parameters and metrics).
+- Reproduce costly experiments (including data, code, models, and pipelines).
+- Organize and compare many experiments.
+
+Over the last few years, experiment tracking tools have come a long way so that
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:32]: > I'd recommend to start the blog post from this - it sets a right mindset for target audience.

👍 

> Also, we can position exp versioning as the combination of these two

Yep, that's what I'm hoping to do. Do you think we should include this diagram?

File: content/blog/2021-11-17-exp-versioning.md [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+@@ -1,5 +1,5 @@
+ train:
+   epochs: 10
+-model:
+-  conv_units: 16
++model:
++  conv_units: 128
+```
+
+Many experiment tracking databases do not capture everything you need to
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:145]: The tricky part is they vary so widely and often support logging arbitrary artifacts that could include anything (although what is realistic to track may be different). My hope was to partially address this with the point in the intro about "codifying pipelines." I'll think about some more. Ideas are welcome!

File: content/blog/2021-11-17-exp-versioning.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+  introduced by each experiment the same way you version code.
+- Codify your experiment pipeline (commands, inputs, and outputs) so you can
+  easily reproduce any experiment.
+- Associate experiments with commits in your repo so you can organize locally
+  and choose what to share.
+
+Experiment versioning builds on modern version control principles and technology
+to address experiment tracking needs.
+
+# Start Experiment Versioning
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:48]: I thought the same and wrote another version where I put this later. However, I struggled to keep it concrete and easy to understand since `exp init` does so much to introduce the project structure and pipeline. For example, how to mention codifying pipelines generically without showing anything DVC-specific? Again, ideas are welcome here!

File: content/blog/2021-11-17-exp-versioning.md [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+- Metrics, plots, and other outputs.
+
+The pipeline has a single `default` stage, but more stages can be added.
+
+With an experiment tracking dashboard, you see all your experiments, and you may
+sort, label, and filter to help keep them organized. With experiment versioning,
+your code and experiments stay connected, and you have more flexibility in how
+you organize and share them.
+
+For example, you may try an experiment in a new Git branch. If something goes
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:186]: Not sure I follow. Do you have something in mind in particular that isn't covered by these two paragraphs, or something that's included but should be in a different paragraph?

File: content/blog/2021-11-17-exp-versioning.md [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+
+This captures:
+
+- The command to run the experiment.
+- Parameters and other dependencies.
+- Metrics, plots, and other outputs.
+
+The pipeline has a single `default` stage, but more stages can be added.
+
+With an experiment tracking dashboard, you see all your experiments, and you may
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:181]: Will it make sense without mentioning `dvc exp show` or DVC specifically? Do we show it again when mentioning how DVC helps? Maybe it's better to remove this sentence and put it in the next section?

File: content/blog/2021-11-17-exp-versioning.md [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+If the experiment instead looks interesting, you can push it along with your
+code so that they stay in sync and are shared with the same people, and it's
+already organized using your existing branch name. You can keep iterating on
+that branch, start a new one if an experiment diverges too much, or merge into
+your main branch to make it your primary model.
+
+# Benefits of DVC
+
+Of course, you could do all this yourself in Git. Since we have established that
+code version control is insufficient for experiments, DVC adds features specific
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-11-17-exp-versioning.md:200]: The benefits you list are pretty narrowly focused on experiment organization and miss other benefits like reproducibility. Do we want to mention things like data management, caching, and pipeline execution that are major benefits of DVC, and are hard to do with either Git or experiment trackers by themselves? Where would you put this?

File: content/blog/2021-11-17-exp-versioning.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (dmpetrov) [reply content/blog/2021-11-17-exp-versioning.md:32]: > Do you think we should include this diagram?

Some visuals would be great not necessary this one.

File: content/blog/2021-11-17-exp-versioning.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (dmpetrov) [reply content/blog/2021-11-17-exp-versioning.md:48]: One good exercise might be to write the entire blog post on idea-level without DVC. Then, add the DVC as a last section. You have partially implemented this - but you can make one step further. Happy to brainstorm this together.

File: content/blog/2021-11-17-exp-versioning.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (dmpetrov) [reply content/blog/2021-11-17-exp-versioning.md:181]: > without mentioning dvc exp show or DVC specifically? 

It might be. You can mention that this is an idea. Or say that you will show later how to create it.

File: content/blog/2021-11-17-exp-versioning.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (dmpetrov) [reply content/blog/2021-11-17-exp-versioning.md:200]: > miss other benefits like reproducibility.

💯 

> data management, caching, and pipeline execution that are major benefits of DVC

not an easy question. It would be great to have these but it will move away the user focus from experiment versioning. It would be great to mention this but not spend too much time on this.

**Comment by dberenbaum** [comment]
@dmpetrov Another version of the blog post is ready.

I left in your ascii drawing, but another option is a tree like we show in the video with experiment leaves growing from the branches.

We may also need to add in some link at the end for where to go to get started.

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+you run and compare them. Pushing an experiment shares all the experiment
+information and code so that they stay in sync and are shared with the same
+people, and it's already organized using Git branches.
+
+```dvc
+$ dvc exp push origin exp-333c9
+Pushed experiment 'exp-333c9'to Git remote 'origin'.
+```
+
+While Git branching is a flexible way to organize experiments, there are often
```
Comment (dmpetrov) [inline content/blog/2021-12-07-ml-experiment-versioning.md:257]: It might worth to write a separate section about ephemeral branches with the link to Peter's blog post about custom references. This is an interesting technical info can attract attention of some folks.

# Experiments as ephemeral Git branches

To keep track of all your experiments you need to support references to all of them. Git branches is an easy way of doing this. However, using Git branches for hundreds of experiments can easily over pollute namespace of your branches - you don't want to see 100s of branches as an output of `git branch` command.

We implemented a special machinery in DVC that we called "ephemeral branches" or experiment references in
order to separate Git branches and experiment branches. This machinery is implemented with [Git 
custom references](https://git-scm.com/book/en/v2/Git-Internals-Git-References).
DVC just introduces a new type of custom references `exps` in addition to the standard ones - `branch` (or `heads`), `remotes`, `tags`.

Git custom references is an advance Git functionality that is used by some services. For example, GitHub stores
pull requests as custom references (proof or code example is needed). DVC uses similar approach to help data scientists and ML engineers manage experiments.

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+
+Experiment versioning is still in its early days. Look out for future
+announcements about:
+
+- Deep learning features like [live monitoring](https://dvc.org/doc/dvclive) and
+  [checkpointing](https://dvc.org/doc/user-guide/experiment-management/checkpoints).
+- Visualizing and comparing experiment results in other tools like VS Code and
+  [DVC Studio](https://studio.iterative.ai/).
+- Orchestrating experiments on other
+  [machines](https://github.com/iterative/dvc/wiki/Remote-executors).
```
Comment (dmpetrov) [inline content/blog/2021-12-07-ml-experiment-versioning.md:288]: It might be a bit too early for `dvc machines` 😄 

Should we point to our Twitter or YouTube?

**Review by dmpetrov — APPROVED** [review]
I like the post more and more 😍 
Some suggestions are inline

I'll try to draft some pictures for it.

File: content/blog/2021-12-07-ml-experiment-versioning.md (hunk 18 above)
Comment (dmpetrov) [reply content/blog/2021-12-07-ml-experiment-versioning.md:257]: The proof with code example: https://stackoverflow.com/questions/70141055/git-fetch-pull-requests-from-github

tl;dr; `git ls-remote origin | grep "refs/pull"`

**Comment by jendefig** [comment]
> I left in your ascii drawing, but another option is a tree like we show in the video with experiment leaves growing from the branches.

I've been wondering about this for a few days.  Might it be better if the blog post image is the tree from the video?  I'm sure we can use the other image we've been working on somewhere (another blog post, presentations, etc.). But it might make more sense because the title of the blog post doesn't exactly tie to the image.  (Tracking is not really what's happening, or versioning - it's just the chaos of it all out of control).

And then per @dmpetrov 's point re: 
>🔥 great section name! is there a way to mention this in the intro\description? 🙂

Maybe add this to the short description, which needs to be as strong as the long description.

"ML Experiments as Code takes ML experiment tracking to the next level; bringing together the benefits of traditional code versioning and modern day experiment tracking, super charging your ability to reproduce your work."

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+  - Experiments
+  - Experiment Tracking
+  - Experiment Versioning
+---
+
+Experiment tracking tools solve problems not handled by version control tools
+like Git. Instead of being built to track only code, they log experiment
+information like parameters, metrics, and even binary artifacts like serialized
+models. They also can organize and compare many experiments, whereas Git may
+become messy and is limited to comparing two versions at a time.
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-12-07-ml-experiment-versioning.md:28]: I'm going to try keeping the first sentence comparing them but then make the rest of the paragraph focus only on experiment tracking. I'm hesitant to drop the comparison completely. It changes the structure of the intro.

With comparison:
1. Where experiment tracking tools beat git.
2. Where git beats experiment tracking.
3. Value of combining them.

Without comparison:
1. Experiment tracking tools are good.
2. Git is good.
3. Why not combine them?
 
Let me know if you think it's better to drop the comparison completely.

File: content/blog/2021-12-07-ml-experiment-versioning.md (hunk 19 above)
Comment (dberenbaum) [reply content/blog/2021-12-07-ml-experiment-versioning.md:288]: > Should we point to our Twitter or YouTube?

@jendefig What do you think?

File: content/blog/2021-12-07-ml-experiment-versioning.md (hunk 18 above)
Comment (dberenbaum) [reply content/blog/2021-12-07-ml-experiment-versioning.md]: Hmm, this is the only comment where I'm not convinced 😄 . This could be interesting technical info for some folks, but does it belong in this post? Technical info isn't the focus of this post, and we can link to Peter's post for those who are interested. It doesn't seem to directly address how experiment versioning improves on experiment tracking (I would argue it's less relevant than concepts like data versioning and pipelines that we are removing). It seems more about how we improve the typical git workflow, which we have generally deprioritized to focus on comparisons to experiment tracking tools. It also doesn't really tie in to other concepts or examples in the post, which is already starting to tilt back towards more abstract and conceptual.

**Comment by dberenbaum** [comment]
@dmpetrov There's another iteration ready if you want to take a look.

**Comment by dberenbaum** [comment]
One other idea from https://github.com/iterative/dvc.org/pull/2909/files#r759706747:

> It might help to discuss why we have this workflow and want to leave changes un-committed. In other experiment tracking tools, the workflow looks like:
>
>    1. Run the experiment.
>    2. Log the commit hash of the current HEAD, even though it doesn't include your dirty changes.
>    3. If everything is successful, save changes in a new commit.
>
> This creates a confusing state where the experiment should really be associated with the second commit instead of the first.
>
> It might be too much detail or inappropriate for the page, but maybe it can be summarized, or it might spark other ideas. This is probably a good point for the blog post but not sure if there's a place for it...

File: content/blog/2021-12-07-ml-experiment-versioning.md (hunk 18 above)
Comment (dmpetrov) [reply content/blog/2021-12-07-ml-experiment-versioning.md]: It's up to you.

I agree that the current version does not fit to the current post. I probably presented it in a wrong way. It should be (if you decide to go with it) more lightweight. Title should be like "Under the hood" or "Git internals". It might be just a simple, easy-to-understand paragraph when we should mention that we use some Git internals, like GitHub does.

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+Path to a model file/directory [models, n to omit]:
+Path to a parameters file [params.yaml, n to omit]:
+Path to a metrics file [metrics.json, n to omit]:
+Path to a plots file/directory [plots, n to omit]: logs.csv
+```
+
+This might mean a little more work to set up your code to read parameters from a
+file or write metrics to another file. To track artifacts like data and models
+in your repo, DVC adds
+[data versioning](https://dvc.org/doc/start/data-and-model-versioning) support.
```
Comment (dmpetrov) [inline content/blog/2021-12-07-ml-experiment-versioning.md:99]: Is it really more work compared to exp tracking. It is (conceptually 🙂 ) the same. It feels like this paragraph can be removed.

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+-  conv_units: 16
++model:
++  conv_units: 128
+```
+
+With DVC, you can compare lots of experiments at a glance:
+
+```dvc
+$ dvc exp show
+┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
```
Comment (dmpetrov) [inline content/blog/2021-12-07-ml-experiment-versioning.md:140]: For me the table is broken a bit
Screen Shot 2021-11-30 at 8 16 52 PM

**Review by dmpetrov — APPROVED** [review]
The post is great we are just polishing it now.

I left a few comments. Mostly in the beginning - the most critical part of the blog post.

PS: If you are thinking about next big improvement steps (which is probably not needed at this point) - we can think about adding more "user's stories". But is it not something we can do fast.

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+tags:
+  - MLOps
+  - DVC
+  - Experiments
+  - Experiment Tracking
+  - Experiment Versioning
+---
+
+Experiment tracking tools help manage machine learning projects where version
+control tools like Git are insufficient. They log parameters and metrics, and
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-12-07-ml-experiment-versioning.md:27]: It looks like we are planning to get the use case merged before this. If we can do that, we could link there, which has a whole table of experiment tracking tools for comparison.

[the remark this answers is not in the document]
Comment (dberenbaum) [reply content/blog/2021-12-07-ml-experiment-versioning.md:27]: cc @jorgeorpinel

File: content/blog/2021-12-07-ml-experiment-versioning.md (hunk 22 above)
Comment (dberenbaum) [reply content/blog/2021-12-07-ml-experiment-versioning.md:140]: Do you mean the "Experiment row" looking not quite aligned? Here's how it looks for me:

Screen Shot 2021-12-01 at 8 49 25 PM

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+ML experiment versioning combines experiment tracking and version control:
+
+- **Experiments as code**: Track meta-information in the code repository and
+  version it like code.
+- **Versioned reproducibility**: Save and restore experiment state, and track
+  changes to only execute what's new.
+- **Distributed experiments**: Organize locally and choose what to share,
+  reusing your existing repo setup.
+
+```
```
Comment (dberenbaum) [inline content/blog/2021-12-07-ml-experiment-versioning.md:49]: @dmpetrov @jendefig @MKhalusova and anyone else, what do you think about leaving in this ascii art vs. removing it?

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 25]
```diff
@@ -0,0 +1,219 @@
...
+-model:
+-  conv_units: 16
++model:
++  conv_units: 128
+```
+
+With DVC, you can even compare lots of experiments from the terminal like you
+would in a dashboard:
+
+```dvctable
```
Comment (shcheklein) [inline content/blog/2021-12-07-ml-experiment-versioning.md:120]: I think `dvctable` doesn't expect the command `$ dvc exp show` at the top ... table should look quite different from what I see below on the screenshots

also, there is a way to colorize the header

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+79/79 [==============================] - 1s 14ms/step - loss: 0.2552 - acc: 0.9180
+Updating lock file 'dvc.lock'
+
+Reproduced experiment(s): exp-be916
+Experiment results have been applied to your workspace.
+
+To promote an experiment to a Git branch run:
+
+        dvc exp branch <exp> <branch>
+
```
Comment (shcheklein) [inline content/blog/2021-12-07-ml-experiment-versioning.md:171]: nit: extra space

File: content/blog/2021-12-07-ml-experiment-versioning.md (hunk 25 above)
Comment (dberenbaum) [reply content/blog/2021-12-07-ml-experiment-versioning.md:120]: It actually works well! Those were before I added `dvctable`. Here it is now:

Screen Shot 2021-12-03 at 5 20 44 PM

File: content/blog/2021-12-07-ml-experiment-versioning.md (hunk 22 above)
Comment (dmpetrov) [reply content/blog/2021-12-07-ml-experiment-versioning.md:122]: I see extra spaces in rows. All the experiment IDs should be one right after another. But I see a space line.

File: content/blog/2021-12-07-ml-experiment-versioning.md [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+--- a/params.yaml
++++ b/params.yaml
+@@ -1,5 +1,5 @@
+ train:
+   epochs: 10
+-model:
+-  conv_units: 16
++model:
++  conv_units: 128
+```
```
Comment (jendefig) [inline content/blog/2021-12-07-ml-experiment-versioning.md:115]: When I did Batuhan's post I learned how to do this. It shows the diffs better:
image 
image

Would that be better?  or does it need to be in DVC?

**Review by jendefig — APPROVED** [review]
